### PR TITLE
YARN scheduler for Heron

### LINKED ISSUE #933: Add scheduler implementation for topology deployment on YARN clusters

### PR DESCRIPTION
**Pull request by ashvina**
YARN scheduler for Heron is implemented using Apache REEF framework. Each Heron topology will be executed as a YARN application, which includes per topology application master and containers. YARN scheduler supports topology submission, kill, and restart. It also provides container HA (currently AM HA is not supported)
Previous related pull requests: https://github.com/twitter/heron/pull/419 and https://github.com/twitter/heron/pull/725

This fixes #933

### CODE REVIEW

**Comment by maosongfu** [comment]
@ashvina Could u resolve the conflicts?

**Comment by ashvina** [comment]
@maosongfu Done. As you may have seen, the conflict did not exist when the pull request was created. It is resolved now. Thanks !

**Comment by kramasamy** [comment]
@maosongfu - let us review this before we get into other conflicts.

**Comment by maosongfu** [comment]
This Pull request looks good to me. 
@ashvina can u add issues to trace remaining TODO in this pull request?
1. Co-locate TMaster and Scheduler
2. Handle the case TMaster fails (ApplicationMaster for YARN job)
3. To avoid uploading the heron-core release package in every topology submission
4. To avoid uploading reef scheduler individually in every topology submission since it is already in heron-core release package.

**Comment by kramasamy** [comment]
@ashvina - please create the issues in github and make a link in the TODO areas in the code.

**Comment by kramasamy** [comment]
@ashvina - In another PR, could you add documentation for REEF scheduler so that we can add a page in the website - this will be very important for people to use.

File: WORKSPACE [hunk 1]
```diff
@@ -192,6 +213,46 @@ maven_jar(
...
+  name = "wake",
+  artifact = "org.apache.reef:wake:" + reef_version
+)
+
+maven_jar(
+  name = "reef-runtime-local",
+  artifact = "org.apache.reef:reef-runtime-local:" + reef_version
+)
+
+maven_jar(
```
Comment (billonahill) [inline WORKSPACE]: please keep deps alphabetized in this file

Comment (ashvina) [inline WORKSPACE]: Fixed. Thanks !

**Comment by kramasamy** [comment]
@maosongfu and @billonahill - if you don't have any other comments, we can merge it. Let me know.

**Pull request merged**